prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+				break
+			}
+			// Pod is running. Do nothing.
+			return nil
+		case corev1.PodRunning:
+			maxPodRunning := r.config().Plank.PodRunningTimeout.Duration
+			if pod.Status.StartTime.IsZero() || time.Since(pod.Status.StartTime.Time) < maxPodRunning {
+				// Pod is still running. Do nothing.
+				return nil
+			}
+
+			// Pod is stuck in running state longer than maxPodRunning
+			// abort the job, and talk to GitHub
+			pj.SetComplete()
+			pj.Status.State = prowv1.AbortedState
+			pj.Status.Description = "Pod running timeout."
+			client, ok := r.buildClients[pj.ClusterAlias()]
+			if !ok {
+				return fmt.Errorf("running pod %s: unknown cluster alias %q", pod.Name, pj.ClusterAlias())
+			}
+			if err := client.Delete(r.ctx, pod); err != nil {
+				return fmt.Errorf("failed to delete pod %s that was in running timeout: %v", pod.Name, err)
+			}
+			r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Deleted stale running pod.")
+		default:
+			// other states, ignore
+			return nil
+		}
+	}
+
+	pj.Status.URL, err = pjutil.JobURL(r.config().Plank, *pj, r.log)
+	if err != nil {
+		r.log.WithFields(pjutil.ProwJobFields(pj)).WithError(err).Warn("failed to get jobURL")
+	}
+
+	if prevPJ.Status.State != pj.Status.State {
+		r.log.WithFields(pjutil.ProwJobFields(pj)).
+			WithField("from", prevPJ.Status.State).
+			WithField("to", pj.Status.State).Info("Transitioning states.")
+	}
+
+	if err := r.pjClient.Patch(r.ctx, pj.DeepCopy(), ctrlruntimeclient.MergeFrom(prevPJ)); err != nil {
+		return fmt.Errorf("patching prowjob: %v", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) syncTriggeredJob(pj *prowv1.ProwJob) (*reconcile.Result, error) {
+	prevPJ := pj.DeepCopy()
+
+	var id, pn string
+
+	pod, podExists, err := r.pod(pj)
+	if err != nil {
+		return nil, err
+	}
+	// We may end up in a state where the pod exists but the prowjob is not
+	// updated to pending if we successfully create a new pod in a previous
+	// sync but the prowjob update fails. Simply ignore creating a new pod
+	// and rerun the prowjob update.
+	if !podExists {
+		// Do not start more jobs than specified and check again later.
+		canExecuteConcurrently, err := r.canExecuteConcurrently(pj)
+		if err != nil {
+			return nil, fmt.Errorf("canExecuteConcurrently: %v", err)
+		}
+		if !canExecuteConcurrently {
+			return &reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		// We haven't started the pod yet. Do so.
+		id, pn, err = r.startPod(pj)
+		if err != nil {
+			if !isRequestError(err) {
+				return nil, fmt.Errorf("error starting pod: %v", err)
+			}
+			pj.Status.State = prowv1.ErrorState
+			pj.SetComplete()
+			pj.Status.Description = "Job cannot be processed."
+			logrus.WithField("job", pj.Spec.Job).WithError(err).Warning("Unprocessable pod.")
+		}
+	} else {
+		id = getPodBuildID(pod)
+		pn = pod.ObjectMeta.Name
+	}
+
+	if pj.Status.State == prowv1.TriggeredState {
```

Comment:
Isn't this a no-op?

Reply:
No, we start off the function with the job being `Triggered` but above this, we might change it to `ErrorState`